bigcommerce/grphp
# PEN-647: Add strategy patterns, h2proxy strategy

## What?

For loading requests into protobuf to hit nghttpx over HTTP1, which then transforms it to h2, to bypass the gRPC C extension entirely in PHP.

nghttpx: https://nghttp2.org/documentation/nghttpx.1.html

This PR:

* Adds a strategy pattern for seamlessly switching between the gRPC and h2proxy strategies for requests while maintaining the public grphp contract and behavior
* Adds a h2proxy strategy with a configuration object that allows for customization, injection, and configuration of the nghttpx connection
* Adds a new Client Request object for passing in request context to strategies
* Adds HeaderRegistry and Header classes for managing collections and instances of headers
* Adds a new Error Status class for cleanly handling error statuses
* Automatically detects if the grpc C extension is not installed, and if not, switches to the h2proxy strategy to allow for seamless timed rollout

## How to Test

Use this config when loading nghttpx: `nghttpx --conf=/path/to/conf`

```
insecure=yes
frontend=0.0.0.0,3000;;no-tls
backend=192.168.33.100,4142;;proto=h2
frontend-frame-debug=yes
log-level=INFO
workers=2
```

Then load up a StoreConfig or A&A `bigcommerce.rpc.Users/GetUser` request from any PHP service via grPHP. It should output the message properly.

From my tests:

Happy path A&A getUser:
```
ID: 1
Email: admin@bigcommerce.com
Array
(
    [content-type] => application/grpc
    [grpc-accept-encoding] => identity,deflate,gzip
    [accept-encoding] => identity,gzip
    [te] => trailers
    [transfer-encoding] => chunked
    [server] => nghttpx
    [via] => h2 linkerd, 2 nghttpx
    [x-proxied-by] => nghttpx
    [grpc-status] => 0
    [grpc-message] => OK
    [l5d-ctx-trace] => +/aqApB9/w+NtTyDUkl7lY21PINSSXuVAAAAAAAAAAA=
    [timer] => 3.548884
    [l5d-success-class] => 1.0
)
```

gRPC error path A&A getUser:

```
Code: user_not_found
Message: User not found with ID 1000
Array
(
    [content-type] => application/grpc
    [grpc-accept-encoding] => identity,deflate,gzip
    [accept-encoding] => identity,gzip
    [te] => trailers
    [transfer-encoding] => chunked
    [server] => nghttpx
    [via] => h2 linkerd, 2 nghttpx
    [x-proxied-by] => nghttpx
    [grpc-status] => 5
    [grpc-message] => User not found with ID 1000
    [error-internal-bin] =>
user_not_foundser not found with ID 1000
    [l5d-success-class] => 1.0
)
```

I've also tested this with the `grpc.so` extension disabled, to ensure it still works then. The `grpc` namespace stubs here are _required_ - because the gRPC client stub code references them in ways I cannot remove, so I need fake stubs to be able to handle it.

----

@bigcommerce/platform-engineering @zackangelo @jmwiese @chrisboulton
